store/types/validity.go

```diff
@@ -1,15 +1,30 @@
 package types
 
+import math "math"
+
+const (
+	// 32K
+	MaxKeyLength = math.MaxUint16
+	// 2G
+	MaxValueLength = math.MaxUint32
```

Comment:
This should cover all cases even in CW?

Reply:
@alpe do you happen to know?

Reply:
This changes a constraint from 4.2GB to 2.1GB. That change isn't safe to make without empirical data.

Reply:
The comment is wrong, the limit is 4G, since it's uint32.

Reply:
There was previously no limit on values, and this PR introduces a limit of MaxUint32 which is 2.1GB.

Reply:
`MaxUint32` is 4G, yeah, it's a breaking change, but it's a necessary change, need to be done sooner or later.

Reply:
That's fine, but breaking changes have enormous and wide-ranging impact, and so must be motivated by something stronger than belief 😃 What concrete rationale supports this breaking change across all Cosmos networks?

Reply:
> That's fine, but breaking changes have enormous and wide-ranging impact, and so must be motivated by something stronger than belief 😃 What concrete rationale supports this breaking change across all Cosmos networks?

deterministic state machine, as I explained in PR description.

Reply:
Maybe instead of making it const, we can make it a variable? make it unbounded by default but bounded for the chains who are sure this is not breaking for them. I think this way you get the best of both worlds?

Reply:
That's a nice goal and makes a lot of things easier in the future, but it isn't possible without breaking users. Example: before this PR, it is possible that some Cosmos node can have state value of size 128GB. This PR clamps that state val to 4GB and therefore breaks that node and its network. Is that acceptable?

Reply:
> Maybe instead of making it const, we can make it a variable? make it unbounded by default but [explitcly] bounded for the chains who are sure this is not breaking for them. I think this way you get the best of both worlds?

sounds good 👍

Reply:
Sorry I am afraid I did not understand, which logic would you like to future proof and how? And what does `external calls` refer to?

Reply:
> Sorry I am afraid I did not understand, which logic would you like to future proof and how? And what does `external calls` refer to?

For example current version of sdk don't use keys larger than 32k, so we can set the 32k limit in our chain safely, but if we don't add the limit inside sdk, there's possibility that future version of sdk use a key larger than 32k, which will break our chain, right?

Reply:
Do you have examples of KV size limits in rocksdb, leveldb, etc? I agree that determinism is needed here, just trying to understand the 4 GiB figure

Reply:
- rocksdb is uint32: [code](https://github.com/facebook/rocksdb/blob/main/db/write_batch.cc#L787)

Reply:
Good call to limit length on k/v for determinism. 👍 
When 2^32 was an implicit hard limit on value length in the supported DBs, then this can be made explicit and covered with tests for regression.

[Writes](https://github.com/cosmos/cosmos-sdk/blob/v0.46.7/store/gaskv/store.go#L52-L53) to the store come with a cost of [30 gas](https://github.com/cosmos/cosmos-sdk/blob/v0.46.7/store/types/gas.go#L238) per bytes already. This should naturally limit the data written. But the max block gas must be set in consensus params. Default in Tendermint is [infinite](https://github.com/informalsystems/tendermint/blob/main/types/params.go#L80) 🙈
 
fyi: I had run some analysis on an uni-5 (juno testnet) dump to get some feeling for length of data used. Max key length was 452 bytes, max val 262144.

@webmaster128 are any limits set on the store for contracts in CosmWasm?

Reply:
Actually we need to be a little bit smaller than hard limit, because iavl leaf node encode both key and value and a few other bytes together as value in low level db, I'd prefer 2G(aka. MaxInt32) if it's enough.

Reply:
Note there is no good reason for having different values in `MAX_LENGTH_DB_KEY` and `MAX_LENGTH_DB_VALUE` in CosmWasm. It was a gut feeling from the early days and if contract developers exceed those they are most likely doing something wrong. For the next iteration of that code, I'd give them both the same value.

Reply:
It's very important that key length is constrained to a value far smaller than value lengths. Key-value stores use keysas  indices into hash map lookup tables, and the efficiency of those lookups depends (first) on the cardinality of the key set.

Reply:
Good to know, thanks @peterbourgon.

In this case my only concern is that the current key limit is smaller than the limit that CosmWasm set before (64 KiB + wasmd prefix + iavl wrapper?) and we have no way to know who is using values close to this limit.

Reply:
CosmWasm limits are subject to constraints established by the SDK, if CosmWasm defines limits that aren't guaranteed to be valid by the SDK which it depends upon those limits aren't valid and are a bug in CosmWasm 🤷

Reply:
Congratulations to this clever analysis. It's not helping though. You asked for empirical data, worried about breaking and other people have been asking for our insights. And yes, in a smart contract scenario we cannot control what people are using right now within the currently enforced range.